src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -991,31 +991,31 @@
                                 :fact_path_id pid
                                 :fact_value_id vid})))
 
-(defn existing-row-ids
+(defn existing-fact-value-ids
   "Returns a map from value to id for each value that's already in the
    named database column."
-  [table column values]
-  (into {}
-        (for [{:keys [value id]}
-              (query-to-vec
-               (format
-                "WITH values AS (SELECT unnest(?) as vals)
-                   SELECT %s AS value, id FROM %s
-                   INNER JOIN values ON values.vals = %s"
-                column table column)
-               (sutils/array-to-param "text" String values))]
-          [value id])))
-
-(defn realize-records!
+  [vhashes]
+  (let [vhashes-param (sutils/array-to-param "bytea" PGobject
+                                             (map sutils/munge-hash-for-storage vhashes))
+        rows (query-to-vec "WITH new_hashes AS (SELECT unnest(?) as h)
+                              SELECT id, encode(value_hash::bytea, 'hex') value_hash FROM fact_values
+                              INNER JOIN new_hashes ON new_hashes.h = value_hash"
```

Comment:
I think it's true that the index won't get used if value_hash is on the RHS, but in practice it doesn't seem to be getting used in either case on the limited set I'm looking at. May be worth digging around to confirm that we can use it.